File: 0733-flood-fill/0733-flood-fill.java
class Solution {
    private void dfs(int[][] image, int row, int col, int from, int to) {
        if (row < 0 || row >= image.length || col < 0 || col >= image[0].length) {
            return;
        }

        if (image[row][col] != from) return;

        image[row][col] = to;

        dfs(image, row - 1, col, from, to);
        dfs(image, row + 1, col, from, to);
        dfs(image, row, col - 1, from, to);
        dfs(image, row, col + 1, from, to);
    }

    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int from = image[sr][sc];

        if (from == color) return image;

        dfs(image, sr, sc, from, color);

        return image;
    }
}

File: 0733-flood-fill/0733-flood-fill.kt
class Solution {
    private fun dfs(image: Array<IntArray>, row: Int, col: Int, from: Int, to: Int) {
        if (row !in image.indices || col !in image[0].indices) {
            return
        }

        if (image[row][col] != from) return

        image[row][col] = to

        dfs(image, row - 1, col, from, to)
        dfs(image, row + 1, col, from, to)
        dfs(image, row, col - 1, from, to)
        dfs(image, row, col + 1, from, to)
    }

    fun floodFill(image: Array<IntArray>, sr: Int, sc: Int, color: Int): Array<IntArray> {
        val from = image[sr][sc]

        if (from == color) return image

        dfs(image, sr, sc, from, color)

        return image
    }
}